sdk/storage/Azure.Storage.Common/src/AsyncCollection.cs

```diff
@@ -0,0 +1,207 @@
+        /// </summary>
+        public virtual int? PageSizeHint { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncCollection{T}"/>
+        /// class for mocking.
+        /// </summary>
+        protected AsyncCollection() =>
+            this.CancellationToken = CancellationToken.None;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncCollection{T}"/>
+        /// class.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// The <see cref="CancellationToken"/> used for requests made while
+        /// enumerating asynchronously.
+        /// </param>
+        protected AsyncCollection(CancellationToken cancellationToken) =>
+            this.CancellationToken = cancellationToken;
+
+        /// <summary>
+        /// Enumerate the values a <see cref="Page{T}"/> at a time.  This may
+        /// make mutliple service requests.
+        /// </summary>
+        /// <param name="continuationToken">
+        /// A continuation token indicating where to resume paging or null to
+        /// begin paging from the beginning.
+        /// </param>
+        /// <returns>
+        /// An async sequence of <see cref="Page{T}"/>s.
+        /// </returns>
+        public abstract IAsyncEnumerable<Page<T>> ByPage(string continuationToken = default);
+
+        /// <summary>
+        /// Enumerate the values in the collection asynchronously.  This may
+        /// make mutliple service requests.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// The <see cref="CancellationToken"/> used for requests made while
+        /// enumerating asynchronously.
+        /// </param>
+        /// <returns>An async sequence of values.</returns>
+        public abstract IAsyncEnumerator<Response<T>> GetAsyncEnumerator(CancellationToken cancellationToken = default);
+
+        /// <summary>
+        /// Enumerate the values in the collection synchronously.  This may
+        /// make mutliple service requests.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// The <see cref="CancellationToken"/> used for requests made while
+        /// enumerating.
+        /// </param>
+        /// <returns>A sequence of values.</returns>
+        protected abstract IEnumerator<Response<T>> GetEnumerator();
+
+        // Explicitly implement these so customers don't accidentally foreach an async enumerable
+        IEnumerator<Response<T>> IEnumerable<Response<T>>.GetEnumerator() => this.GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
+
+    }
+
+    /// <summary>
+    /// A single <see cref="Page{T}"/> of values from a request that may return
+    /// zero or more <see cref="Page{T}"/>s of values.
+    /// </summary>
+    /// <typeparam name="T">The type of values.</typeparam>
+    public readonly struct Page<T> : IEquatable<Page<T>>
```

Comment:
I don't think we should make it equatable. Still hide the Equals(object) though

Reply:
Yes - Removed per Pavel/Keith's suggestions above.